File: src/Shadowing.py
#!/usr/bin/env python3

import numpy as np

def intersect_AB_t(A, B, t):
  """
  Intersection of line AB with triangle t, 3-dimensional.

  Reference: https://www.iue.tuwien.ac.at/phd/ertl/node114.html
  Note: p = A + B*x, not (B-A)*x

  """

  EPS = 0.0
  area = np.zeros(3)

  for i in range(3):
    C = np.cross(t[i%3,:]-A, t[(i+1)%3,:]-A)
    area[i] = 0.5*np.dot(B, C)

  if area[0] >= -EPS and area[1] >= -EPS and area[2] >= -EPS:
    has_solution = True
  elif area[0] <= +EPS and area[1] <= +EPS and area[2] <= +EPS:
    has_solution = True
  else:
    has_solution = False

  if has_solution:
    C = np.zeros(3)
    area /= np.sum(area)
    for i in range(3):
      C = C + area[i]*t[i,:]

  return C, has_solution

def non(mu_i, mu_e):
  """Non-illuminated || non-visible won't be computed."""

  nu_i = np.ones(len(mu_i))
  nu_e = np.ones(len(mu_e))

  for i in range(len(mu_i)):
    if mu_i[i] == 0.0 or mu_e[i] == 0.0:
      nu_i[i] = 0.0
      nu_e[i] = 0.0

  return nu_i, nu_e

def nu(faces, nodes, normals, centres, s, nu_i):
  """Shadowing, non-convex version."""

  t = np.zeros((3,3))

  for i in range(len(faces)):

    if nu_i[i] > 0.0:

      A = centres[i,:]
      B = s
 
      j = 0
      while j < len(faces) and nu_i[i] > 0.0:
        if i != j:
          if nu_i[j] > 0.0:
 
            C = centres[j,:] - centres[i,:]
            tmp = np.dot(C, normals[i,:])
            if tmp > 0.0:
 
              for k in range(3):
                t[k,:] = nodes[faces[j,k],:]
 
              C, has_solution = intersect_AB_t(A, B, t)
 
              if has_solution:
                nu_i[i] = 0.0
        j += 1

  return nu_i

def main():
  pass

if __name__ == "__main__":
  main()

File: tvet/fortran/shadowing.f90
! shadowing.f90
! Shadowing, non-convex version.
! Miroslav Broz (miroslav.broz@email.cz), Oct 26th 2022

module shadowing_module

contains

! Dicrectional cosine.

subroutine mu(normals, s, mu_i)

implicit none
double precision, dimension(:,:), intent(in) :: normals
double precision, dimension(3), intent(in) :: s
double precision, dimension(:), intent(out) :: mu_i

integer :: i

!$omp parallel do private(i) shared(normals,mu_i,s)
do i = 1, size(normals,1)
  mu_i(i) = max(dot_product(normals(i,:), s), 0.d0)
enddo
!$omp end parallel do

return
end subroutine mu

! Non-illuminated || non-visible won't be computed.

subroutine non(mu_i, mu_e, nu_i, nu_e)

implicit none
double precision, dimension(:), intent(in) :: mu_i, mu_e
double precision, dimension(:), intent(out) :: nu_i, nu_e

integer :: i

nu_i = 1.d0
nu_e = 1.d0
!$omp parallel do private(i) shared(mu_i,mu_e,nu_i,nu_e)
do i = 1, size(mu_i,1)
  if ((mu_i(i).eq.0.d0).or.(mu_e(i).eq.0.d0)) then
    nu_i(i) = 0.d0
    nu_e(i) = 0.d0
  endif
enddo
!$omp end parallel do

return
end subroutine non

! Mutual shadowing.

subroutine nu(faces, nodes, normals, centres, s, nu_i)

use boundingbox_module
use intersect_AB_t_module

implicit none
integer, dimension(:,:), intent(in) :: faces
double precision, dimension(:,:), intent(in) :: nodes, normals, centres
double precision, dimension(3), intent(in) :: s
double precision, dimension(:), intent(inout) :: nu_i

integer :: i, j, k
double precision, dimension(3,3) :: t
double precision, dimension(3) :: A, B, C
double precision, dimension(3) :: area
double precision :: tmp
logical :: has_solution

double precision, dimension(:,:), pointer, save :: boxes

call boundingbox(faces, nodes, s, boxes)

!$omp parallel do private(i,j,k,A,B,C,t,has_solution,tmp,area) shared(nodes,faces,centres,s,nu_i)
do i = 1, size(faces,1)
  if (nu_i(i).gt.0.d0) then

    A = centres(i,:)
    B = s

    j = 1
    do while ((j.le.size(faces,1)).and.(nu_i(i).gt.0.d0))

      if (i.ne.j) then
        if (nu_i(j).gt.0.d0) then

          if ((boxes(j,2).lt.boxes(i,1)).or.(boxes(j,1).gt.boxes(i,2) &
            .or.(boxes(j,4).lt.boxes(i,3)).or.(boxes(j,3).gt.boxes(i,4)))) then
            j = j+1
            cycle
          endif

          C = centres(j,:)-centres(i,:)
          tmp = dot_product(C, normals(i,:))
          if (tmp.gt.0.d0) then

            do k = 1, 3
              t(k,:) = nodes(faces(j,k),:)
            enddo

            call intersect_AB_t(A, B, t, C, has_solution)

            if (has_solution) then
              nu_i(i) = 0.d0
            endif

          endif

        endif
      endif

      j = j+1
    enddo  ! j

  endif
enddo ! i
!$omp end parallel do

return
end subroutine nu

end module shadowing_module


